### PR TITLE
[REF] EventNormalizer: complete refactoring

### PR DESCRIPTION
**Pull request by Gorash**

### CODE REVIEW

File: packages/core/src/EventNormalizer.ts [hunk 1]
```diff
@@ -146,326 +536,812 @@ export class EventNormalizer {
...
+            (cutEvent && 'deleteByCut') ||
+            (dropEvent && 'insertFromDrop') ||
+            (pasteEvent && 'insertFromPaste') ||
+            (isGoogleKeyboardBackspace && 'deleteContentBackward') ||
+            (!!googleKeyboardKey && 'insertText') ||
+            (inputEvent && inputEvent.inputType) ||
+            // todo: Do we really need to set the inputType when making a
+            //       "special accent" in mac?
+            (isAccentMac && 'insertCompositionText');
+        const defaultPrevented =
```
Comment (Gorash) [inline packages/core/src/EventNormalizer.ts]: never used. Why ? It's important to allow the prevent default by plugins and external libs

File: examples/getKeys/index.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<html>
```
[the remark this answers is not in the document]
Comment (Goaman) [reply examples/getKeys/index.html]: it is not polished now, we can move it in a different branch before merging

File: packages/core/src/EventManager.ts [hunk 3]
```diff
@@ -23,20 +23,28 @@ export class EventManager {
...
-                return ['insertParagraphBreak', {}];
+                if (action.text === '\n') {
+                    return ['insertLineBreak', {}];
+                } else {
+                    return ['insertText', { text: action.text }];
+                }
             case 'selectAll':
             case 'setSelection': {
+                // Todo: This is wrong in case the normalizer provide a range in the DOM that is not
+                //       present in the VDocument.
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/src/EventManager.ts]: note: `parser.parseSelection` bug

File: packages/core/test/eventNormalizerUtils.ts [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    textContent?: string;
+    targetId?: number;
+    removedNodes?: RemovedNodesTargetMutation[];
+    addedNodes?: AddedNodesTargetMutation[];
+}
+
+export interface TestSelectionEvent {
+    type: 'selection';
+    focus: {
+        targetSelectionId: number;
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: It's the "id" of the node to focus provided by `NodeIndexContainer`. Do you want me to rename it by `id`?

File: packages/core/test/eventNormalizerUtils.ts [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    let type = 'unknown';
+    Object.keys(types).forEach(function(key) {
+        const isType = types[key].some(function(str) {
+            return eventName.toLowerCase().indexOf(str) !== -1;
+        });
+        if (isType) {
+            type = key;
+        }
+    });
+    return type;
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: The function was only moved in this file without modifications. But your argument is still good. I have changed it.

File: packages/core/test/eventNormalizerUtils.ts [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+}
+class InputEventBis extends InputEvent {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: the class was only moved from `eventNormalizer.ts`

File: packages/core/test/eventNormalizerUtils.ts [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+        selection.addRange(domRange);
+    }
+    selection.extend(endContainer, endOffset);
+}
+
+export async function nextTick(): Promise<void> {
+    return await new Promise((resolve): number => setTimeout(resolve));
+}
+
+export class NodeIndexContainer {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: I have documented the class

File: packages/core/test/eventNormalizerUtils.ts [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+ *
+ * Encode the key of the offsetCacheMap with "<parentId>$<childIndex>" (e.g. if
+ * the parent id is 'a' and the offset is 0, the key will be encoded as "a$0").
+ */
+export async function triggerEvents(eventStackList: TestEvent[][]): Promise<void> {
+    const addedNodes: Node[] = [];
+    const editableElement = document.getElementById('editable');
+    const nodeIndexContainer = new NodeIndexContainer(editableElement);
+    for (const eventStack of eventStackList) {
+        eventStack.forEach((testEvent): void => {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: ?

File: packages/core/test/eventNormalizerUtils.ts [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+            if (testEvent.type === 'mutation') {
+                const mutationEvent = testEvent;
+                const targetId = mutationEvent.targetId;
+                const mutatedNode = nodeIndexContainer.getNode(targetId);
+                if (testEvent.mutationType === 'childList') {
+                    if (testEvent.removedNodes) {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: what kind of util? don't get the reason?

File: packages/core/test/eventNormalizerUtils.ts [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+                        });
+                    }
+                } else if (mutationEvent.mutationType === 'characterData') {
+                    mutatedNode.textContent = mutationEvent.textContent;
+                }
+            } else if (testEvent.type === 'selection') {
+                if (testEvent.anchor.targetSelectionId) {
+                    const selectionEvent = testEvent;
+                    setRange(
+                        nodeIndexContainer.getNode(selectionEvent.anchor.targetSelectionId),
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: yes, the `selectionEvent.anchor.targetSelectionId` is an id recorder with a `nodeIndexContainer`

File: packages/core/test/eventNormalizerUtils.ts [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+                        selectionEvent.focus.offset,
+                    );
+                }
+            } else {
+                const { type, ...options } = testEvent;
+                triggerEvent(editableElement, type, options as TriggerNativeEventsOption);
+            }
+        });
+        await nextTick();
+        await nextTick();
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: I have commented it
```
        // After the first nextTick(which trigger a setTimeout), there is still
        // some microtasks that might execute.  We wait another tick (i.e.
        // setTimeout) to be sure every microtasks have been executed.
```

File: packages/core/test/eventNormalizerUtils.ts [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+    }
+}
+
+export interface TestContext {
+    container: HTMLElement;
+    editable: HTMLElement;
+    div: HTMLElement;
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: That element is used in `EventNormalizerPointer.test.ts`. I haven't touched the logic that used it.

File: packages/core/test/eventNormalizerUtils.ts [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+    eventBatches: EventBatch[];
+    normalizer: EventNormalizer;
+}
+
+export function testCallbackBefore(): TestContext {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/test/eventNormalizerUtils.ts]: The file `EventNormalizer.test.ts` have been splited in two `EventNormalizerKeyoard.test.ts` and `EventNormalizerPointer.test.ts`. This function is used in both files now.

File: packages/plugin-devtools/test/devtools.test.ts [hunk 14]
```diff
@@ -738,15 +738,15 @@ describe('Plugin: DevTools', () => {
...
+            container.dispatchEvent(new InputEvent('input', { bubbles: true }));
             await nextTickFrame();
             await nextTickFrame();
 
             const button = devtools.querySelector('devtools-navbar button:not(.selected)');
             await click(button);
 
             const subpanel = devtools.querySelector('devtools-panel.active mainpane-contents');
-            expect(subpanel.textContent).to.equal('insertTextsetSelection');
+            expect(subpanel.textContent).to.equal('insertTextsetSelectionsetSelection');
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-devtools/test/devtools.test.ts]:
```
nameBold.dispatchEvent(
                new MouseEvent('dblclick', { clientX: pos.left, clientY: pos.top, bubbles: true }),
            );
```
trigger the fist 'setSelection' while the keydown interpreted by the normalizer trigger the second one.

File: packages/core/src/EventNormalizer.ts [hunk 15]
```diff
@@ -334,460 +625,476 @@ export class EventNormalizer {
...
-     * @param {KeyboardEvent} keypress
-     */
-    _enrichByKeypress(ev: CompiledEvent, keypress: KeyboardEvent): void {
-        if (keypress.key !== 'Dead') {
-            ev.key = keypress.key;
+        if (cutEvent) {
+            const deleteContentAction: DeleteContentAction = {
+                type: 'deleteContent',
+                direction: Direction.FORWARD,
+            };
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/src/EventNormalizer.ts]: The cut event is triggered only if in the editable: 
```typescript
        this._bindEvent(editable, 'cut', this._onClipboard);
```

File: packages/core/src/EventNormalizer.ts [hunk 16]
```diff
@@ -334,460 +625,476 @@ export class EventNormalizer {
...
+            normalizedActions = [deleteContentAction];
+        } else if (dropEvent) {
+            normalizedActions.push(...this._getDropActions(dropEvent));
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/src/EventNormalizer.ts]: yes, `_getDropActions` can provide whether  `[DeleteContentAction, SetSelectionAction]` or ` [SetSelectionAction]`.

File: packages/core/src/EventNormalizer.ts [hunk 17]
```diff
@@ -334,460 +625,476 @@ export class EventNormalizer {
...
+            inputEvent &&
+            inputEvent.inputType &&
+            inputEvent.inputType.indexOf('format') === 0
+        ) {
+            const formatName = inputEvent.inputType.replace('format', '').toLowerCase();
+
+            const applyFormatAction: ApplyFormatAction = {
+                type: 'applyFormat',
+                format: formatName,
+                data: inputEvent.data,
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/src/EventNormalizer.ts]: In w3c (https://www.w3.org/TR/input-events-2/#interface-InputEvent-Attributes) depending on the format, the data is different:
> data holds information plaintext data related to what is to be added to the document.

File: packages/utils/polyfill.ts [hunk 18]
```diff
@@ -1,25 +1,33 @@
...
-            offset: 0,
+            offsetNode: carretRange.startContainer,
+            offset: carretRange.startOffset,
         };
     }
-    return caretPosition;
+
+    // Target deepest node.
+    let offsetNode = caretPosition.offsetNode;
+    let offset = caretPosition.offset;
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/utils/polyfill.ts]: done

File: packages/utils/src/Dom.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+ * @param container
+ * @param offset
+ */
+export function targetDeepest(container: Node, offset: number): [Node, number] {
+    while (container.hasChildNodes()) {
+        if (offset >= container.childNodes.length) {
+            container = container.lastChild;
+            // The new container might be a text node, so considering only
+            // the `childNodes` property would be wrong.
+            offset = nodeLength(container);
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/utils/src/Dom.ts]: I changed code to call this function in caretPositionFromPoint

File: packages/core/src/EventManager.ts [hunk 20]
```diff
@@ -35,22 +35,24 @@ export class EventManager {
                     return ['insertText', { text: action.text }];
                 }
             case 'selectAll':
+                // todo: add selectAll command in the editor
```
Comment (dmo-odoo) [inline packages/core/src/EventManager.ts]: Why the TODO ? The command *does* exist.

File: packages/core/src/EventNormalizer.ts [hunk 21]
```diff
@@ -464,19 +394,33 @@ export class EventNormalizer {
...
         const document = this.editable.ownerDocument;
         this._bindEvent(editable, 'compositionstart', this._registerEvent);
         this._bindEvent(editable, 'compositionupdate', this._registerEvent);
         this._bindEvent(editable, 'compositionend', this._registerEvent);
         this._bindEvent(editable, 'beforeinput', this._registerEvent);
         this._bindEvent(editable, 'input', this._registerEvent);
 
+        this._bindEvent(document, 'onkeydown', this._onSelectionChange);
+        this._bindEvent(document, 'onkeypress', this._onSelectionChange);
+        this._bindEvent(document, 'onkeyup', this._onSelectionChange);
```
Comment (dmo-odoo) [inline packages/core/src/EventNormalizer.ts]: ?

File: packages/core/src/EventNormalizer.ts [hunk 22]
```diff
@@ -485,6 +429,8 @@ export class EventNormalizer {
         this._bindEvent(document, 'touchstart', this._onPointerDown);
         this._bindEvent(editable, 'keydown', this._onKeyDownOrKeyPress);
         this._bindEvent(editable, 'keypress', this._onKeyDownOrKeyPress);
+        this._bindEvent(document, 'onkeyup', this._updateModifiersKeys);
```
Comment (dmo-odoo) [inline packages/core/src/EventNormalizer.ts]: ?

File: packages/core/src/EventNormalizer.ts [hunk 23]
```diff
@@ -1517,24 +1383,25 @@ export class EventNormalizer {
...
      * Catch setSelection and selectAll actions
      *
-     * @private
      * @param {MouseEvent} ev
      */
     _onContextMenu(ev: MouseEvent): void {
-        clearTimeout(this._clickTimeout);
-        this._clickTimeout = setTimeout(() => {
+        window.clearTimeout(this._clickTimeout);
+        this._clickTimeout = window.setTimeout(() => {
```
Comment (dmo-odoo) [inline packages/core/src/EventNormalizer.ts]: This `setTimeout` isn't in a promise, is it expected ?

File: packages/core/src/EventNormalizer.ts [hunk 24]
```diff
@@ -1590,10 +1455,9 @@ export class EventNormalizer {
         // When the users clicks in the DOM, the range is set in the next tick.
         // The observation of the resulting range must thus be delayed to the
         // next tick as well. Store the data we have now before it gets invalid.
-        this._initialCaretPosition = this._locateEvent(ev);
+        this._initialCaretPosition = this._getEventCaretPosition(ev);
 
-        // clearTimeout(this._clickTimeout);
-        this._clickTimeout = setTimeout(() => this._analyzeSelectionChange(ev), 0);
+        this._clickTimeout = window.setTimeout(() => this._analyzeSelectionChange(ev), 0);
```
Comment (dmo-odoo) [inline packages/core/src/EventNormalizer.ts]: This one is a `setTimeout` outside of a promise as well.

**Comment by dmo-odoo** [comment]
🎉

**Pull request merged**